rclcpp/include/rclcpp/parameter_client.hpp

```diff
@@ -73,28 +74,30 @@ class AsyncParametersClient
     std::function<void(
       std::shared_future<std::vector<rclcpp::parameter::ParameterVariant>>)> callback = nullptr)
   {
-    std::promise<std::vector<rclcpp::parameter::ParameterVariant>> promise_result;
-    auto future_result = promise_result.get_future().share();
+    auto promise_result = std::make_shared<std::promise<std::vector<rclcpp::parameter::ParameterVariant>>>();
+    get_parameters_promises_.insert(promise_result);
+    auto future_result = promise_result->get_future().share();
 
     auto request = std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
     request->names = names;
 
     get_parameters_client_->async_send_request(
       request,
-      [&names, &promise_result, &future_result, &callback](
+      [names, promise_result, &future_result, &callback, this](
         rclcpp::client::Client<rcl_interfaces::srv::GetParameters>::SharedFuture cb_f) {
           std::vector<rclcpp::parameter::ParameterVariant> parameter_variants;
           auto & pvalues = cb_f.get()->values;
 
-          std::transform(pvalues.begin(), pvalues.end(), std::back_inserter(parameter_variants),
-          [&names, &pvalues](rcl_interfaces::msg::ParameterValue pvalue) {
+          for(auto & pvalue : pvalues) {
             auto i = &pvalue - &pvalues[0];
             rcl_interfaces::msg::Parameter parameter;
             parameter.name = names[i];
             parameter.value = pvalue;
-            return rclcpp::parameter::ParameterVariant::from_parameter(parameter);
-          });
-          promise_result.set_value(parameter_variants);
+            parameter_variants.push_back(rclcpp::parameter::ParameterVariant::from_parameter(parameter));
+          }
+
+          promise_result->set_value(parameter_variants);
+          get_parameters_promises_.erase(promise_result);
```

Comment:
Can you prefix this with `this->`? I had trouble matching the use of `this` in the lambda to a specific line.